src/prometheus/prometheus.c

```diff
@@ -0,0 +1,358 @@
+  else if (strcasecmp(key, "timeout") == 0)
+    return config_set(&timeout, value);
+  else if (strcasecmp(key, "socket") == 0)
+    return config_set(&sock, value);
+  else if (strcasecmp(key, "jwttoken") == 0)
+    return config_set(&jwt_token, value);
+  else
+    return -1;
+}
+
+static int prometheus_init(void) {
+  if (curl != NULL)
+    curl_easy_cleanup(curl);
+
+  if ((curl = curl_easy_init()) == NULL) {
+    ERROR("prometheus plugin: curl_easy_init failed.");
+    return -1;
+  }
+
+  curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
+  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, prometheus_write_callback);
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, COLLECTD_USERAGENT);
+  curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, prometheus_curl_error);
+
+  if (user != NULL) {
+#ifdef HAVE_CURLOPT_USERNAME
+    curl_easy_setopt(curl, CURLOPT_USERNAME, user);
+    curl_easy_setopt(curl, CURLOPT_PASSWORD, (pass == NULL) ? "" : pass);
+#else
+    static char credentials[1024];
+    int status = ssnprintf(credentials, sizeof(credentials), "%s:%s", user,
+                           pass == NULL ? "" : pass);
+    if ((status < 0) || ((size_t)status >= sizeof(credentials))) {
+      ERROR("prometheus plugin: Credentials would have been truncated.");
+      return -1;
+    }
+
+    curl_easy_setopt(curl, CURLOPT_USERPWD, credentials);
+#endif
+  }
+
+  curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
+  curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 50L);
+
+  if ((verify_peer == NULL) || IS_TRUE(verify_peer)) {
+    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1L);
+  } else {
+    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
+  }
+
+  if ((verify_host == NULL) || IS_TRUE(verify_host)) {
+    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 2L);
+  } else {
+    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
+  }
+
+  if (cacert != NULL) {
+    curl_easy_setopt(curl, CURLOPT_CAINFO, cacert);
+  }
+
+#ifdef HAVE_CURLOPT_TIMEOUT_MS
+  if (timeout != NULL) {
+    curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, atol(timeout));
+  } else {
+    curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS,
+                     (long)CDTIME_T_TO_MS(plugin_get_interval()));
+  }
+#endif
+
+#ifdef HAVE_CURLOPT_UNIX_SOCKET_PATH
+  if (sock != NULL) {
+    curl_easy_setopt(curl, CURLOPT_UNIX_SOCKET_PATH, sock);
+  }
+#endif
+
+  static char jwt_header[512];
+  int status = ssnprintf(jwt_header, sizeof(jwt_header),
+                         "Authorization: Bearer %s", jwt_token);
+  if ((status < 0) || ((size_t)status >= sizeof(jwt_header))) {
+    ERROR("prometheus plugin: jwt token would have been truncated.");
+    return -1;
+  }
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER,
+                   curl_slist_append(NULL, jwt_header));
+
+  return 0;
+}
+
+static metric_t *convert_pr_metric_to_metric(pr_metric_t *pr_metric,
+                                             metric_type_t metric_type) {
+  metric_t *metric = malloc(sizeof(*metric));
```

Comment:
Allocs here and elsewhere in code could use collectd `scalloc()`, `smalloc()` and `sstrdup()` helpers which log an error and exit on allocation failures.

(There's rarely any point in continuing after alloc fails, as functionality on such error conditions is rarely tested to work correctly, and rest of system is pretty hosed at that point.)